File: rb101/rb101-rb109-small_problems/easy8/all_substrings.rb
def substrings_at_start(string)
  substrings = []
  string.chars.each_index do |idx|
    substrings << string[0..idx]
  end
  substrings
end

def substrings(string)
  all_substrings = []
  string.chars.each_index do |idx|
    all_substrings.concat(substrings_at_start(string[idx..-1]))
  end
  all_substrings
end

p substrings('abcde') == [
  'a', 'ab', 'abc', 'abcd', 'abcde',
  'b', 'bc', 'bcd', 'bcde',
  'c', 'cd', 'cde',
  'd', 'de',
  'e'
]

# ALTERNATIVE SOLUTIONS:

# 1)
# def substrings(string)
#   results = []
#   (0...string.size).each do |start_index|
#     this_substring = string[start_index..-1]
#     results.concat(substrings_at_start(this_substring))
#   end
#   results
# end

# 2)
# def substrings(string)
#   all_substrings = []
#   string.chars.each_index do |idx|
#     all_substrings << substrings_at_start(string[idx..-1])
#   end
#   all_substrings.flatten
# end

File: js210/js210-small_problems/list_processing/allSubstrings.js
function substrings(string) {
  let substrings = [];
  for (let index = 0; index < string.length; index += 1) {
    substrings = substrings.concat(leadingSubstrings(string.slice(index)));
  }

  return substrings;
}

function leadingSubstrings(string) {
  let substrings = [];
  for (let length = 1; length <= string.length; length += 1) {
    substrings.push(string.slice(0, length));
  }

  return substrings;
}

console.log(substrings('abcde'));

// returns (as shown by console output):
// [ "a", "ab", "abc", "abcd", "abcde",
//   "b", "bc", "bcd", "bcde",
//   "c", "cd", "cde",
//   "d", "de",
//   "e" ]


// ALTERNATIVE SOLUTION:

// function substrings(string) {
//   let substrings = [];
//   for (let startIndex = 0; startIndex < string.length; startIndex += 1) {
//     let substring = string.substring(startIndex);
//     substrings = substrings.concat(leadingSubstrings(substring));
//   }

//   return substrings;
// }

// function leadingSubstrings(string) {
//   let substrings = [];
//   for (let length = 1; length <= string.length; length += 1) {
//     substrings.push(string.slice(0, length));
//   }

//   return substrings;
// }


// FURTHER EXPLORATION:

// 1)
// function substrings(string) {
//   let letters = string.split('');

//   return letters.map((_letter, idx) => {
//     return leadingSubstrings(letters.slice(idx).join(''));
//   }).reduce((acc, curArr) => acc.concat(curArr));
// }

// 2)
// function substrings(string) {
//   return string.split('').map((_letter, idx) => {
//     return leadingSubstrings(string.slice(idx));
//   }).flat();
// }
